rockets/jsonrpc/receiver.cpp

```diff
@@ -153,45 +157,91 @@ class Receiver::Impl
         if (!_isValidJsonRpcRequest(request))
         {
             if (isNotification)
-                callback(json());
+                respond(json());
             else
-                callback(makeErrorResponse(invalidRequest, id));
+                respond(makeErrorResponse(invalidRequest, id));
             return;
         }
 
         const auto methodName = request["method"].get<std::string>();
+
+        if(methodName == cancelMethodName)
+        {
+            processCancel(request, respond);
+            return;
+        }
+
         const auto method = methods.find(methodName);
         if (method == methods.end())
         {
             if (isNotification)
-                callback(json());
+                respond(json());
             else
-                callback(makeErrorResponse(methodNotFound, id));
+                respond(makeErrorResponse(methodNotFound, id));
             return;
         }
 
+        const auto& cancelFunc = cancels.find(methodName);
+        if(cancelFunc != cancels.end())
+            pendingRequests[id] = {cancelFunc->second, respond};
+
         const auto params = request.find("params") == request.end()
                                 ? ""
                                 : dump(request["params"]);
         const auto& func = method->second;
         func({params, clientID},
-             [callback, id](const Response rep) {
+             [respond, id, &pendingRequests = pendingRequests](const Response rep) {
                  // No reply for valid "notifications" (requests without an
                  // "id")
                  if (id.is_null())
-                     callback(json());
+                     respond(json());
                  else
-                     callback(makeResponse(rep, id));
+                     respond(makeResponse(rep, id));
+                 pendingRequests.erase(id);
              });
     }
+
+    void processCancel(const json& request, JsonResponseCallback respond)
+    {
+        // need a valid notification with the ID of the request to cancel
+        const auto id = request.count("id") ? request["id"] : json();
+        const bool isNotification = id.is_null();
+        if(!isNotification || request.find("params") == request.end() ||
+           !request["params"].count("id") )
+        {
+            respond(json());
+            return;
+        }
+
+        // invalid request ID or request already processed
+        const auto requestID = request["params"]["id"];
+        auto cancelFunc = pendingRequests.find(requestID);
+        if(cancelFunc == pendingRequests.end())
+        {
+            respond(json());
+            return;
+        }
+
+        // cancel callback to the application
+        cancelFunc->second.first();
+
+        // respond to request
+        cancelFunc->second.second(makeErrorResponse(requestAborted, requestID));
+
+        pendingRequests.erase(requestID);
+
+        respond(json());
```

Comment:
respond(json()); can be done outside of this function, after calling processCancel(), since it never responds anything

Reply:
good point